File: HongJun/Tasks/Libraries/ODE/testScript/ode45Tests.py
import os, sys, unittest, math
import numpy as np
from testScript import read_tests

currentdir = os.path.dirname(os.path.realpath(__file__))
parentdir = os.path.dirname(currentdir)
sys.path.append(parentdir)

from ode import ode45

class Testode45(unittest.TestCase):
    
    '''This script is used to perform unit tests for the ode45 function. This script 
        will read the text file produced by the ode45Tests.m file, it will use the 
        inputs to test the ode45 function and then compare the actual results with the
        expected results. Note: any new functions added to the MATLAB script also have 
        to be added to this script.
    '''
    
    def test_ode45(self):
        
        results,inputs=read_tests("ode45.txt",True)
        
        for i in range(len(inputs)):
            inp = inputs[i]
            res = results[i]
            
            if inp.fun == 'polyNN':
                self.polyNN(res)
                pass
            elif inp.fun == 'ball':
                self.ball(res)
                pass
            elif inp.fun == 'trigbasic':
                self.trigevents(res)
                pass
            elif inp.fun == 'trigbasic2':
                self.trigtol(res)
                pass
            elif inp.fun == 'cosbasic1':
                self.cosbasic1(res)
                pass
            elif inp.fun == 'cosbasic2':
                self.cosbasic2(res)
                pass
            elif inp.fun == 'cosbasic3':
                self.cosbasic3(res)
                pass
            else:
                print("Function not recognized : "+str(inp.fun))
            
        

    def polyNN(self,result):
        tspan = [-6,5]
        y0=[25,50,25,50]
        opt={'NonNegative':[0,1]}
        
        def f(t,y):
            x=0.02*(3*pow(t,5)-62*pow(t,3)+42*pow(t,2)+45*t+18)
            return [x,x,x,x]
        
        sol = ode45(f,tspan,y0,opt)
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
        
    
    def ball(self,result):
        y0 = [0,20]
        tspan = [0,30]
        
        def f(t,y):
            return [y[1], -9.8]
        
        def events(t,y):
            return [y[0]],[1],[-1]
        
        opt = {'Events':events}
        sol = ode45(f,tspan,y0,opt)
        teout,yeout,ieout=sol.get_events()
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
        self.compare_events(result,teout,yeout,ieout)
            
    
    def trigevents(self,result):
        y0 = [0,-1]
        tspan = [0,30]
        
        def f(t,y):
            return [np.cos(t),np.sin(t)]
        
        def events(t,y):
            return [y[0],y[1],y[0]],[0,0,0],[-1,0,1]
        
        opt = {'Events':events}
        sol = ode45(f,tspan,y0,opt)
        teout,yeout,ieout=sol.get_events()
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
        self.compare_events(result,teout,yeout,ieout)
    
    
    def trigtol(self,result):
        def f(t,y):
            return [y[0]*np.cos(t),y[1]*np.sin(t)]    

        opt={'RelTol' : 0.0009,'AbsTol' : 3.e-8,'NormControl' : 'on','Refine' : 3,'NonNegative' : [0, 1]}
        
        Tspan = [19, 53]
        Y0 = [6, 38]
        sol=ode45(f,Tspan,Y0,opt)
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
    
    
    def cosbasic1(self,result):
        def f(t,y):
            return [math.cos(t),2*math.cos(t)]    
        
        mass = np.array([[0.46,0.16],[6.4,0.11]])
        opt={'Mass' : mass, 'NonNegative' : [0]}
        
        Tspan = [8, 59]
        Y0 =[84, 0.9]
        sol=ode45(f,Tspan,Y0,opt)
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
    
    
    
    def cosbasic2(self,result):
        
        def f(t,y):
            return [math.cos(t),2*math.cos(t)]

        opt={'AbsTol':[1e-2,2e-7],'RelTol':1e-6}
        tspan = [8000000000000,8000000000010]
        y0 = [1, 0]
        sol=ode45(f,tspan,y0,opt)
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
        
        
    def cosbasic3(self,result):
        
        def f(t,y):
            return [math.cos(t),2*math.cos(t)]


        opt={'AbsTol':[1e-2,2e-7],'RelTol':1e-6,'InitialStep':0.2}
        tspan = np.linspace(0,10,21)
        y0 = [1, 0]
        sol=ode45(f,tspan,y0,opt)
        self.compare_ty(result,sol.tout,sol.yout)
        self.compare_stats(result,sol.get_stats())
    
    
    
    def compare_ty(self,res,tout,yout):
        
        t = res.tout
        y = res.yout
        
        self.assertEqual(len(t),len(tout))
        self.assertEqual(len(y),len(yout)*len(yout[0]))
        
        for i in range(len(t)):
            for j in range(len(yout)):
                self.assertAlmostEqual(y[j*len(t)+i],yout[j][i])
            self.assertAlmostEqual(tout[i],t[i])
    
    def compare_stats(self,res,statsvec):
        self.assertEqual(res.nsteps,statsvec[0])
        self.assertEqual(res.nfailed,statsvec[1])
        self.assertEqual(res.nfevals,statsvec[2])
       
        
    def compare_events(self,res,teout,yeout,ieout):
        te=res.teout
        ye=res.yeout
        ie=res.ieout
        
        self.assertEqual(len(te),len(teout))
        self.assertEqual(len(ye),len(yeout)*len(yeout[0]))
        self.assertEqual(len(ie),len(ieout))
        
        for i in range(len(te)):
            for j in range(len(yeout)):
                self.assertAlmostEqual(ye[j*len(te)+i],yeout[j][i])
            self.assertAlmostEqual(teout[i],te[i])
            self.assertEqual(ieout[i],ie[i]-1)
        


if __name__ == "__main__":
    unittest.main()
    

File: HongJun/Tasks/Libraries/ODE/testScript/ode45Tests.m

function main()
    %This script is used to perform unit tests for the ode45 function. This script 
    % will generate the expected output for a series of chosen inputs, it will then 
    % write these inputs and expected outputs into a text file. More tests can be 
    % added if desired (Note: any new tests will also have to be added to the Python script)
    
    fileID = fopen('ode45.txt','w');
    
    opt=odeset('NonNegative',[1,2]);
    tspan = [-6,5];
    y0=[25,50,25,50];
    [t,y]=ode45(@polyNN,tspan,y0,opt);
    sol=ode45(@polyNN,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,[],[],[]);
    
    
    y0 = [0;20];
    tspan = [0,30];
    opt = odeset('Events',@ballevents);
    [t,y,te,ye,ie] = ode45(@ball,tspan,y0,opt);
    sol=ode45(@ball,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,te,ye,ie);

    
    y0 = [0;-1];
    tspan = [0,30];
    opt = odeset('Events',@trigevents);
    [t,y,te,ye,ie] = ode45(@trigbasic,tspan,y0,opt);
    sol=ode45(@trigbasic,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,te,ye,ie);
    
    
    y0 = [6; 38];
    tspan = [19, 53];
    opt=odeset('RelTol' , 0.0009,'AbsTol' , 3.e-5,'NormControl' , 'on','Refine' , 3,'NonNegative' , [1, 2]);
    [t,y]=ode45(@trigbasic2,tspan,y0,opt);
    sol=ode45(@trigbasic2,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,[],[],[]);

    
    y0 = [84; 0.9];
    tspan = [8, 59];
    mass = [[0.46,0.16];[6.4,0.11]];
    opt=odeset('Mass',mass,'NonNegative' , [1]);
    [t,y]=ode45(@cosbasic1,tspan,y0,opt);
    sol=ode45(@cosbasic1,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,[],[],[]);
    
    
    y0 = [1; 0];
    tspan = [8000000000000,8000000000010];
    opt=odeset('AbsTol',[1e-2,2e-7],'RelTol',1e-6);
    [t,y]=ode45(@cosbasic2,tspan,y0,opt);
    sol=ode45(@cosbasic2,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,[],[],[]);
    
    
    y0 = [1; 0];
    tspan = [0:0.5:10];
    opt=odeset('AbsTol',[1e-2,2e-7],'RelTol',1e-6,'InitialStep',0.2);
    [t,y]=ode45(@cosbasic3,tspan,y0,opt);
    sol=ode45(@cosbasic3,tspan,y0,opt);
    writetext(fileID,sol,tspan,y0,t,y,[],[],[]);
    
end


%% Write all your functions here, odeFcn, Events, Mass

function dydt = polyNN(t,y)
    dydt = [0.02*(3*t^5-62*t^3+42*t^2+45*t+18);0.02*(3*t^5-62*t^3+42*t^2+45*t+18);0.02*(3*t^5-62*t^3+42*t^2+45*t+18);0.02*(3*t^5-62*t^3+42*t^2+45*t+18)];
end

function dydt = ball(t,y)
dydt = [y(2); -9.8];
end

function dydt = trigbasic(t,y)
    dydt = [cos(t);sin(t)];
end

function dydt = trigbasic2(t,y)
    dydt = [y(1)*cos(t);y(2)*sin(t)];
end

function dydt = cosbasic1(t,y)
    dydt=[cos(t);2*cos(t)];
end

function dydt = cosbasic2(t,y)
    dydt=[cos(t);2*cos(t)];
end

function dydt = cosbasic3(t,y)
    dydt=[cos(t);2*cos(t)];
end

function [value,isterminal,direction] = ballevents(t,y)
value = [y(1)];
isterminal = [1];
direction = [-1];
end

function [value,isterminal,direction] = trigevents(t,y)
value = [y(1),y(2),y(1)];
isterminal = [0,0,0];
direction = [-1,0,1];
end


%%

function writetext(fileID,sol,tspan,y0,t,y,te,ye,ie)
    
    %Inputs
    fprintf(fileID,'Function:');
    fprintf(fileID,'%s',func2str(sol.extdata.odefun));
    fprintf(fileID,' ');
    fprintf(fileID,'Tspan:');
    fprintf(fileID,'%.15f#',tspan);
    fprintf(fileID,' ');
    fprintf(fileID,'Y0:');
    fprintf(fileID,'%.15f#',y0);
    fprintf(fileID,' ');
    fprintf(fileID,'Varargin:');
    fprintf(fileID,'%.15f#',string(sol.extdata.varargin));
    fprintf(fileID,' ');
    
    %Options
    fprintf(fileID,'RelTol:');
    fprintf(fileID,'%.15f',sol.extdata.options.RelTol);
    fprintf(fileID,' ');
    fprintf(fileID,'AbsTol:');
    fprintf(fileID,'%.15f#',sol.extdata.options.AbsTol);
    fprintf(fileID,' ');
    fprintf(fileID,'NormControl:');
    fprintf(fileID,'%s',sol.extdata.options.NormControl);
    fprintf(fileID,' ');
    fprintf(fileID,'Refine:');
    fprintf(fileID,'%d',sol.extdata.options.Refine);
    fprintf(fileID,' ');
    fprintf(fileID,'Stats:');
    fprintf(fileID,'%s',sol.extdata.options.Stats);
    fprintf(fileID,' ');
    fprintf(fileID,'NonNegative:');
    fprintf(fileID,'%d#',sol.extdata.options.NonNegative);
    fprintf(fileID,' ');
    fprintf(fileID,'Events:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%s',func2str(sol.extdata.options.Events));
    end
    fprintf(fileID,' ');
    fprintf(fileID,'MaxStep:');
    fprintf(fileID,'%.15f',sol.extdata.options.MaxStep);
    fprintf(fileID,' ');
    fprintf(fileID,'InitialStep:');
    fprintf(fileID,'%.15f',sol.extdata.options.InitialStep);
    fprintf(fileID,' ');
    fprintf(fileID,'Mass:');
    if isa(sol.extdata.options.Mass,'function_handle')
        fprintf(fileID,'%s',func2str(sol.extdata.options.Mass));
    else
        fprintf(fileID,'%.15f#',sol.extdata.options.Mass);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'MStateDependence:');
    fprintf(fileID,'%s',sol.extdata.options.MStateDependence);
    fprintf(fileID,' ');

    %Output
    fprintf(fileID,'Tout:');
    fprintf(fileID,'%.15f#',t);
    fprintf(fileID,' ');
    fprintf(fileID,'Yout:');
    fprintf(fileID,'%.15f#',y);
    fprintf(fileID,' ');
    fprintf(fileID,'Nsteps:');
    fprintf(fileID,'%d',sol.stats.nsteps);
    fprintf(fileID,' ');
    fprintf(fileID,'Nfailed:');
    fprintf(fileID,'%d',sol.stats.nfailed);
    fprintf(fileID,' ');
    fprintf(fileID,'Nfevals:');
    fprintf(fileID,'%d',sol.stats.nfevals);
    fprintf(fileID,' ');
    fprintf(fileID,'Teout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%.15f#',te);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'Yeout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%.15f#',ye);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'Ieout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%d#',ie);
    end
    fprintf(fileID,'\n');
    
end